warehouse/manage/views.py

```diff
@@ -616,6 +616,70 @@ def delete_webauthn(self):
         return HTTPSeeOther(self.request.route_path("manage.account"))
 
 
+@view_defaults(
+    uses_session=True,
+    require_csrf=True,
+    require_methods=False,
+    permission="manage:user",
+    http_cache=0,
+    has_translations=True,
+)
+class ProvisionRecoveryCodesViews:
+    def __init__(self, request):
+        self.request = request
+        self.user_service = request.find_service(IUserService, context=None)
+
+    @view_config(
+        request_method="GET",
+        route_name="manage.account.recovery-codes.generate",
+        renderer="manage/account/recovery_codes-provision.html",
+    )
+    def recovery_codes_generate(self):
+        if self.user_service.has_recovery_codes(self.request.user.id):
+            self.request.session.flash(
+                _("Recovery codes already generated"), queue="error"
+            )
+            return HTTPSeeOther(self.request.route_path("manage.account"))
+
+        self.user_service.record_event(
+            self.request.user.id,
+            tag="account:recovery_codes:generated",
+            ip_address=self.request.remote_addr,
+        )
+        return {
+            "recovery_codes": self.user_service.generate_recovery_codes(
+                self.request.user.id
+            )
+        }
```

Comment:
This allows the user to generate recovery codes without having any 2FA provisioned -- should it? If not, the button to generate codes should probably be disabled as well.

Reply:
put a guard in the view in https://github.com/pypa/warehouse/pull/7260/commits/199adf0ca51dbb6fbaf488bb2a15b61ae3f57eaa, I'll hide the button in the template as well in next commit.